internal/factory/container/device.go

```diff
@@ -176,12 +176,53 @@ func (c *container) specAddContainerConfigDevices(enableDeviceOwnershipFromSecur
 }
 
 func (c *container) specInjectCDIDevices() error {
-	// TODO: Once CRI is extended with native CDI support this will need to be updated...
-	_, names, err := cdi.ParseAnnotations(c.Config().GetAnnotations())
+	var (
+		cdiDevices = c.Config().CDIDevices
+		fromCRI    = map[string]struct{}{}
+		requested  = make([]string, 0, len(cdiDevices))
+		annotated  []string
+		err        error
+	)
+
+	// Take CDI devices from the dedicated CDIDevices CRI field.
+	for _, dev := range cdiDevices {
+		requested = append(requested, dev.Name)
+		fromCRI[dev.Name] = struct{}{}
+	}
+
+	// Extract CDI devices from annotations which is still supported as a means
+	// of injecting CDI devices to give people time to update their DRA drivers.
+	// TODO(klihub): Change the log message to a warning once annotations are
+	// deprecated, and to an error once support is removed altogether.
+	_, annotated, err = cdi.ParseAnnotations(c.Config().GetAnnotations())
```

Comment:
should we mark them as deprecated for 1.28 and then remove them in 1.29?

Reply:
Yes, I think we should mark it deprecated. About removing support for it altogether, I think we should go in sync with what @bart0sh is doing on that front in containerd.